source/server/listener_manager_impl.cc

```diff
@@ -410,56 +427,129 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
     // If we got here and found already configured branch, then it means that this FilterChainMatch
     // is a duplicate, and that there is some overlap in the repeated fields with already processed
     // FilterChainMatches.
     throw EnvoyException(fmt::format("error adding listener '{}': multiple filter chains with "
                                      "overlapping matching rules are defined",
                                      address_->asString()));
   }
-  source_types_array[source_type] = filter_chain;
 }
 
-void ListenerImpl::convertDestinationIPsMapToTrie() {
+namespace {
+
+// Template function for creating a CIDR list entry for either source or destination address.
+template <class T>
+std::pair<T, std::vector<Network::Address::CidrRange>> makeCidrListEntry(const std::string& cidr,
+                                                                         const T& data) {
+  std::vector<Network::Address::CidrRange> subnets;
+  if (cidr == EMPTY_STRING) {
+    if (Network::Address::ipFamilySupported(AF_INET)) {
+      static const std::string catch_all_ipv4 = "0.0.0.0/0";
+      subnets.push_back(Network::Address::CidrRange::create(catch_all_ipv4));
+    }
+    if (Network::Address::ipFamilySupported(AF_INET6)) {
+      static const std::string catch_all_ipv6 = "::/0";
+      subnets.push_back(Network::Address::CidrRange::create(catch_all_ipv6));
+    }
+  } else {
+    subnets.push_back(Network::Address::CidrRange::create(cidr));
+  }
+  return std::make_pair<T, std::vector<Network::Address::CidrRange>>(T(data), std::move(subnets));
+}
+
+}; // namespace
+
+void ListenerImpl::convertIPsToTries() {
   for (auto& port : destination_ports_map_) {
+    // These variables are used as we build up the destination CIDRs used for the trie.
     auto& destination_ips_pair = port.second;
     auto& destination_ips_map = destination_ips_pair.first;
-    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>> list;
+    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>>
+        destination_ips_list;
+    destination_ips_list.reserve(destination_ips_map.size());
+
     for (const auto& entry : destination_ips_map) {
-      std::vector<Network::Address::CidrRange> subnets;
-      if (entry.first == EMPTY_STRING) {
-        if (Network::Address::ipFamilySupported(AF_INET)) {
-          subnets.push_back(Network::Address::CidrRange::create("0.0.0.0/0"));
-        }
-        if (Network::Address::ipFamilySupported(AF_INET6)) {
-          subnets.push_back(Network::Address::CidrRange::create("::/0"));
+      destination_ips_list.push_back(makeCidrListEntry(entry.first, entry.second));
+
+      // This hugely nested for loop greatly pains me, but I'm not sure how to make it better.
+      // We need to get access to all of the source IP strings so that we can convert them into
+      // a trie like we did for the destination IPs above.
```

Comment:
Can't we do this more locally, once we've recursed down to source address levels?

Reply:
How do you mean? Like break it out into a different function?

Reply:
I was thinking that we could build up the trie as we built up the non-optimized version, but it's probably even more complicated than this two phase approach. Fine to keep as is.